File: src/main/java/algo/prefixSum/BinarySubarraysWithSum_930.java
package algo.prefixSum;

public class BinarySubarraysWithSum_930 {
  public static int numSubarraysWithSum(int[] nums, int goal) {
    var len = nums.length;
    var sum = new int[len + 1];
    for (var i = 0; i < len; i++) {
      sum[i + 1] = sum[i] + nums[i];
    }

    var count = 0;
    for (var i = 0; i <= len; i++) {
      for (var j = i + 1; j <= len; j++) {
        if (sum[j] - sum[i] == goal) {
          count++;
        } else if (sum[j] - sum[i] > goal) {
          break;
        }
      }
    }
    return count;
  }

  public static void main(String[] args) {
    System.out.println(numSubarraysWithSum(new int[]{1,0,1,0,1}, 2));
    System.out.println(numSubarraysWithSum(new int[]{0,0,0,0,0}, 0));
  }
}

File: src/main/kotlin/algo/prefixSum/BinarySubarraysWithSum_930.kt
package algo.prefixSum

class KtBinarySubarraysWithSum_930 {
    fun numSubarraysWithSum(nums: IntArray, goal: Int): Int {
        val sums: MutableMap<Int, Int> = mutableMapOf()  // sum -> number of sums
        var sum = 0
        var count = 0
        nums.indices.forEach { idx ->
            val num = nums[idx]
            sum += num
            if (sum == goal) {
                count++
            }
            sums[sum - goal]?.also { count += it }
            sums[sum] = sums.getOrDefault(sum, 0) + 1
        }
        return count
    }
}

fun main() {
    val subarray = KtBinarySubarraysWithSum_930()
    println(subarray.numSubarraysWithSum(intArrayOf(1,0,1,0,1), 2))
    println(subarray.numSubarraysWithSum(intArrayOf(0,0,0,0,0), 0))
    println(subarray.numSubarraysWithSum(intArrayOf(0,1,1,1,1), 3))
    println(subarray.numSubarraysWithSum(intArrayOf(0,0,0,0,0,0,1,0,0,0), 0))
}